File: app-ios/tutanotaTests/CompatibilityTest.m
#import <XCTest/XCTest.h>
#import "TUTEncodingConverter.h"
#import <Foundation/Foundation.h>
#import "Swiftier.h"


@interface CompatibilityTest : XCTestCase

@property NSDictionary *testData;

@end

@implementation CompatibilityTest

- (void)setUp {
    [super setUp];
	NSBundle *bundle = [NSBundle bundleForClass:[self class]];
	NSString *path =  [bundle pathForResource:@"CompatibilityTestData" ofType:@"json"];
    NSData *jsonData = [[NSData alloc] initWithContentsOfFile:path];
	NSError *error = nil;
	self.testData = [NSJSONSerialization JSONObjectWithData:jsonData options:NSJSONReadingAllowFragments error:&error];

	if (error){
		XCTFail(@"error %@", error);
	}
}

- (void)testUnicodeEncoding {
	for (NSDictionary *td in self.testData[@"encodingTests"]) {
		NSData *encoded = [TUTEncodingConverter stringToBytes:td[@"string"]];
		XCTAssertEqualObjects([TUTEncodingConverter base64ToBytes:td[@"encodedString"]], encoded);
		XCTAssertEqualObjects([TUTEncodingConverter bytesToBase64:encoded], td[@"encodedString"]);
		NSString *decoded = [TUTEncodingConverter bytesToString:encoded];
		XCTAssertEqualObjects(decoded, td[@"string"]);
		return;
	}
}

- (void)testEncodingSimple {
	NSData *data = [TUTEncodingConverter stringToBytes:@"abc"];
	NSData* b64Data = [TUTEncodingConverter base64ToBytes:@"YWJj"];
	NSData* hexData = [TUTEncodingConverter hexToBytes:@"616263"];

	XCTAssertEqualObjects(@"abc", [TUTEncodingConverter bytesToString:data]);
	XCTAssertEqualObjects(@"abc", [TUTEncodingConverter bytesToString:b64Data]);
	XCTAssertEqualObjects(@"abc", [TUTEncodingConverter bytesToString:hexData]);

	XCTAssertEqualObjects(@"YWJj", [TUTEncodingConverter bytesToBase64:data]);
	XCTAssertEqualObjects(@"YWJj", [TUTEncodingConverter bytesToBase64:b64Data]);
	XCTAssertEqualObjects(@"YWJj", [TUTEncodingConverter bytesToBase64:hexData]);


	XCTAssertEqualObjects(@"abc", [TUTEncodingConverter bytesToString:data]);
	XCTAssertEqualObjects(@"abc", [TUTEncodingConverter bytesToString:b64Data]);
	XCTAssertEqualObjects(@"abc", [TUTEncodingConverter bytesToString:hexData]);

	XCTAssertEqualObjects(@"616263", [TUTEncodingConverter bytesToHex:data]);
	XCTAssertEqualObjects(@"616263", [TUTEncodingConverter bytesToHex:hexData]);
	XCTAssertEqualObjects(@"616263", [TUTEncodingConverter bytesToHex:hexData]);
}


- (void)testEncodingEmptyString {
	NSData *data = [TUTEncodingConverter stringToBytes:@""];
	NSData* b64Data = [TUTEncodingConverter base64ToBytes:@""];
	NSData* hexData = [TUTEncodingConverter hexToBytes:@""];

	XCTAssertEqualObjects(@"", [TUTEncodingConverter bytesToString:data]);
	XCTAssertEqualObjects(@"", [TUTEncodingConverter bytesToString:b64Data]);
	XCTAssertEqualObjects(@"", [TUTEncodingConverter bytesToString:hexData]);

	XCTAssertEqualObjects(@"", [TUTEncodingConverter bytesToBase64:data]);
	XCTAssertEqualObjects(@"", [TUTEncodingConverter bytesToBase64:b64Data]);
	XCTAssertEqualObjects(@"", [TUTEncodingConverter bytesToBase64:hexData]);


	XCTAssertEqualObjects(@"", [TUTEncodingConverter bytesToString:data]);
	XCTAssertEqualObjects(@"", [TUTEncodingConverter bytesToString:b64Data]);
	XCTAssertEqualObjects(@"", [TUTEncodingConverter bytesToString:hexData]);

	XCTAssertEqualObjects(@"", [TUTEncodingConverter bytesToHex:data]);
	XCTAssertEqualObjects(@"", [TUTEncodingConverter bytesToHex:hexData]);
	XCTAssertEqualObjects(@"", [TUTEncodingConverter bytesToHex:hexData]);
}

@end

File: app-ios/tutanotaTests/CompatibilityTest.swift
import Testing
import tutanota

struct EncodingTest: Decodable {
	let string: String
	let encodedString: String
}

struct TestData: Decodable {
	let encodingTests: [EncodingTest]
}

class CompatibilityTest {
	private let testData: TestData
	init() throws {
		let bundle = Bundle(for: Self.self)
		let dataurl = bundle.url(forResource: "CompatibilityTestData", withExtension: "json")!
		let jsonData = try Data(contentsOf: dataurl)
		testData = try JSONDecoder().decode(TestData.self, from: jsonData)
	}

	@Test func testUnicodeEncoding() {
		for td in self.testData.encodingTests {
			let encoded = TUTEncodingConverter.string(toBytes: td.string)
			#expect(TUTEncodingConverter.base64(toBytes: td.encodedString) == encoded)
			#expect(TUTEncodingConverter.bytes(toBase64: encoded) == td.encodedString)

			let decoded = TUTEncodingConverter.bytes(toString: encoded)
			#expect(decoded == td.string)
		}
	}

	@Test func testUnicodeEncodingSwift() {
		for td in self.testData.encodingTests {
			let encoded = td.string.data(using: .utf8)!
			#expect(Data(base64Encoded: td.encodedString)!  == encoded)
			#expect(encoded.base64EncodedString() == td.encodedString)

			let decoded = String(data: encoded, encoding: .utf8)!
			#expect(decoded == td.string)
		}
	}

	// FIXME the rest?

	//	@Test func testEncodingSimple() {
	//		let data = TUTEncodingConverter.string(toBytes: "abc")
	//		let b64Data = TUTEncodingConverter.base64(toBytes: "YWJj")
	//	}
	//}
}
